src/main/core.js

```diff
@@ -66,41 +68,108 @@ function coreFormat(text, opts, addAlignmentSize) {
-    if (cursorNode) {
-      cursorOffset = opts.cursorOffset - opts.locStart(cursorNode);
-      opts.cursorNode = cursorNode;
+    const nodeResult = rangeUtil.findNodeAtOffset(ast, opts.cursorOffset, opts);
+    if (nodeResult && nodeResult.node) {
+      opts.cursorNode = nodeResult.node;
     }
   }
 
   const astComments = attachComments(text, ast, opts);
   const doc = printAstToDoc(ast, opts, addAlignmentSize);
   opts.newLine = guessLineEnding(text);
 
-  const result = printDocToString(doc, opts);
-  const formatted = result.formatted;
-  const cursorOffsetResult = result.cursor;
+  let result = printDocToString(doc, opts);
+
   ensureAllCommentsPrinted(astComments);
   // Remove extra leading indentation as well as the added indentation after last newline
   if (addAlignmentSize > 0) {
-    return { formatted: formatted.trim() + opts.newLine };
+    const trimmed = result.formatted.trim();
+    const cursorNodeStart =
+      result.cursorNodeStart === undefined
+        ? undefined
+        : result.cursorNodeStart - result.formatted.indexOf(trimmed);
+    result = Object.assign({}, result, {
+      formatted: trimmed + opts.newLine,
+      cursorNodeStart
+    });
   }
 
-  if (cursorOffset !== undefined) {
+  if (opts.cursorOffset >= 0) {
+    let oldCursorNodeStart;
+    let oldCursorNodeText;
+
+    let cursorOffsetRelativeToOldCursorNode;
+
+    let newCursorNodeStart;
+    let newCursorNodeText;
+
+    if (opts.cursorNode && result.cursorNodeText) {
+      oldCursorNodeStart = opts.locStart(opts.cursorNode);
+      oldCursorNodeText = text.slice(
+        oldCursorNodeStart,
+        opts.locEnd(opts.cursorNode)
+      );
+
+      cursorOffsetRelativeToOldCursorNode =
+        opts.cursorOffset - oldCursorNodeStart;
+
+      newCursorNodeStart = result.cursorNodeStart;
+      newCursorNodeText = result.cursorNodeText;
+    } else {
+      oldCursorNodeStart = 0;
+      oldCursorNodeText = text;
+
+      cursorOffsetRelativeToOldCursorNode = opts.cursorOffset;
+
+      newCursorNodeStart = 0;
+      newCursorNodeText = result.formatted;
+    }
+
+    if (oldCursorNodeText === newCursorNodeText) {
+      return {
+        formatted: result.formatted,
+        cursorOffset: newCursorNodeStart + cursorOffsetRelativeToOldCursorNode
+      };
+    }
+
+    // diff old and new cursor node texts, with a special cursor
+    // symbol inserted to find out where it moves to
+
+    const CURSOR = Symbol("cursor");
+
+    const oldCursorNodeCharArray = oldCursorNodeText.split("");
+    oldCursorNodeCharArray.splice(
+      cursorOffsetRelativeToOldCursorNode,
+      0,
+      CURSOR
+    );
+
+    const newCursorNodeCharArray = newCursorNodeText.split("");
+
+    const cursorNodeDiff = diff.diffArrays(
+      oldCursorNodeCharArray,
+      newCursorNodeCharArray
+    );
+
+    let i = 0;
+    for (const entry of cursorNodeDiff) {
+      if (entry.removed) {
+        if (entry.value.indexOf(CURSOR) > -1) {
+          break;
```

Comment:
Maybe?

```js
if (entry.removed && entry.value.indexOf(CURSOR) > -1) {
```

Reply:
It's not the same thing, because there's an `else` clause. Would need to change the `else` to `else if (!entry.removed)` in order for that to work, which IMO is a bit more confusing. But happy to change it if you prefer.

Reply:
Oops sorry!